File: v1/src/main/java/com/api/v1/customer/utils/CustomerFinderUtil.java
package com.api.v1.customer.utils;

import com.api.v1.customer.domain.Customer;
import com.api.v1.customer.domain.CustomerRepository;
import com.api.v1.customer.exceptions.CustomerNotFoundException;
import com.api.v1.person.domain.PersonRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
public class CustomerFinderUtil {

    @Autowired
    private PersonRepository personRepository;

    @Autowired
    private CustomerRepository customerRepository;

    public Mono<Customer> find(String ssn) {
        return personRepository
                .findBySsn(ssn)
                .switchIfEmpty(Mono.error(CustomerNotFoundException::new))
                .flatMap(person -> customerRepository.findByPerson(person));
    }

}

File: kotlin/v3/src/main/kotlin/com/api/v3/customers/utils/CustomerFinderUtil.kt
package com.api.v3.customers.utils

import com.api.v3.customers.domain.Customer
import com.api.v3.customers.domain.CustomerRepository
import com.api.v3.customers.exceptions.CustomerNotFoundException
import com.api.v3.persons.utils.PersonFinderUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.single
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class CustomerFinderUtil {

    @Autowired
    private lateinit var personFinderUtil: PersonFinderUtil

    @Autowired
    private lateinit var customerRepository: CustomerRepository

    suspend fun find(ssn: String): Customer {
        return withContext(Dispatchers.IO) {
            val existingPerson = personFinderUtil.find(ssn)
            if (existingPerson == null) {
                throw CustomerNotFoundException()
            }
            customerRepository
                .findAll()
                .filter { e -> e.person == existingPerson }
                .single()
        }
    }

}